### PR TITLE
Enable object as parameter for css, prop, and attr assertions

### LINKED ISSUE #7353: Passing an object to 'css', 'prop', or 'attr' assertions always passes
### Current behavior:

Passing an object as the first parameter to the '`css`, `prop`, or `attr` assertions will always pass, regardless of the element's actual properties.

### Desired behavior:

There are two possible courses of action here:
- Throw an error when an object is passed
- Modify the functionality to accept an object (in addition to the current behavior) and check for all key/value pairs within that object.

It appears that the second course of action is preferred, as it will make it simpler when performing multiple assertions and be more consistent with how a developer may expect it to function.

### Test code to reproduce

```javascript
this.$div = $('<div style=\'display: none\'>div</div>')

// all will pass
expect(this.$div).to.have.css({ display: 'none' })
expect(this.$div).to.have.css({ display: 'inline' })
expect(this.$div).to.have.css({ foo: 'bar' })

// all will fail
expect(this.$div).not.to.have.css({ display: 'none' })
expect(this.$div).not.to.have.css({ display: 'inline' })
expect(this.$div).not.to.have.css({ foo: 'bar' })
```

This same behavior can be observed with `prop` and `attr`.

### Versions

4.5.0

### PR DESCRIPTION
**Pull request by panzarino**
- Closes #7353

### User facing changelog

Now have the option to pass an object of values to the `attr`, `prop`, and `css` assertion methods.

### Additional details

This change was needed because previously passing an object would always make the assertion pass. An exception could be thrown, but actually being able to pass an object would be a better option.

### How has the user experience changed?

Take the following test for example:
```javascript
this.$div = $('<div style=\'display: none; position: absolute;\'>div</div>')

// Group 1 - should pass
expect(this.$div).to.have.css({ display: 'none' })
expect(this.$div).to.have.css({ position: 'absolute' })
expect(this.$div).to.have.css({ display: 'none', position: 'absolute' })

// Group 2 - should not pass
expect(this.$div).to.have.css({ display: 'inline' })
expect(this.$div).to.have.css({ position: 'fixed' })
expect(this.$div).to.have.css({ display: 'inline', position: 'fixed' })
expect(this.$div).to.have.css({ foo: 'bar' })

// Group 3 - should pass
expect(this.$div).not.to.have.css({ display: 'inline' })
expect(this.$div).not.to.have.css({ position: 'fixed' })
expect(this.$div).not.to.have.css({ display: 'inline', position: 'fixed' })
expect(this.$div).not.to.have.css({ foo: 'bar' })
```

The intended behavior is that groups 1 and 3 pass, while group 2 does not.

Before these changes, groups 1 and 2 would pass while group 3 would not. The screenshot also shows that the correct comparison was not being done:
Screen Shot 2020-05-14 at 8 36 45 PM

After these changes, the correct comparisons are done and group 1 passes while group 2 does not. The screenshot shows how the correct comparison is used:
Screen Shot 2020-05-14 at 8 38 41 PM

As intended, groups 1 and 3 both pass with the correct comparisons being done:
Screen Shot 2020-05-14 at 8 40 12 PM

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue been tagged with a release in ZenHub? 
- [x] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [x] Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?

Docs changes: https://github.com/cypress-io/cypress-documentation/pull/2803

### CODE REVIEW

File: packages/driver/test/cypress/integration/commands/assertions_spec.js [hunk 1]
```diff
@@ -2350,7 +2350,7 @@ describe('src/cy/commands/assertions', () => {
 
     context('attr', () => {
       beforeEach(function () {
-        this.$div = $('<div foo=\'bar\'>foo</div>')
+        this.$div = $('<div foo=\'bar\' fizz=\'buzz\'>foo</div>')
```
Comment (jennifer-shehane) [inline packages/driver/test/cypress/integration/commands/assertions_spec.js:2353]: You can use double quotes here - these escaped single quote `\'` are just an artifact of converting from CoffeeScript. I'd prefer double quotes.

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]
### 1. Assertion display differences

The display of the assertions differs when you use Cypress `.should()` assertions versus the standard expect. When using `.should()` it always only displays the last assertion in the Command Log. You should be able to write tests around the number/value of logs displayed in this situation.

My code against `<div style="display: block; position: absolute;">div</div>`

```js
it('visit', () => {
  cy.visit('index.html')
  cy.get('div').then(($div) => {
    expect($div).to.have.css({ display: 'block' })
    expect($div).to.have.css({ position: 'absolute' })
    expect($div).to.have.css({ display: 'block', position: 'absolute' })
  })
})

it('visit', () => {
  cy.visit('index.html')
  cy.get('div').should('have.css', { display: 'block' })
  cy.get('div').should('have.css', { position: 'absolute' })
  cy.get('div').should('have.css', { display: 'block', position: 'absolute' })
})
```

Screen Shot 2020-05-20 at 12 36 54 PM

### 2. Empty object behavior

The assertions always pass when passed an empty Object. I dunno, maybe this is ok? Seems a bit weird though. The assertions don't display in the UI at all, so it seems like a noop.

```js
it('visit', () => {
  cy.visit('index.html')
  cy.get('div').then(($div) => {
    expect($div).to.have.css({})
  })
})

it('visit', () => {
  cy.visit('index.html')
  cy.get('div').should('have.css', {})
})
```

Screen Shot 2020-05-20 at 12 41 42 PM

File: cli/types/index.d.ts [hunk 2]
```diff
@@ -1,33 +1,33 @@
...
+// This way jQuery can load it using 'reference types="sizzle"' directive
+
+// load ambient declaration for "cypress" NPM module
+// hmm, how to load it better?
+/// <reference path="./cypress-npm-api.d.ts" />
+
+/// <reference path="./cypress.d.ts" />
+/// <reference path="./cypress-global-vars.d.ts" />
+/// <reference path="./cypress-type-helpers.d.ts" />
+/// <reference path="./cypress-expect.d.ts" />
```
Comment (Bkucera) [inline cli/types/index.d.ts:33]:
```suggestion
/// <reference path="./cypress-expect.d.ts" />

```

**Review by kuceb — CHANGES_REQUESTED** [review]

**Comment by panzarino** [comment]
@chrisbreiding and I decided that its probably best to not make these changes as they are fairly inconsistent with other methods and the existing Chai api, plus there would have to be a unique workaround to fix the logging issues. For now, I'm just going to submit another PR that causes a test to fail when passed an object rather than blindly succeed.

**Comment by brian-mann** [comment]
I am not convinced this should not be implemented... I don't believe there are logging issues with this.

**Comment by jennifer-shehane** [comment]
Perhaps this illustrates the issue more clearly. When using `.should()`, only the last assertion in the object is logged as opposed to logging all assertions within a normal expect.

@brian-mann Can you explain more clearly why you consider this not an issue?

Screen Shot 2020-06-12 at 11 27 49 AM

Also, my second point about asserting on an empty object hasn't been responded to - whether this is desired behavior or not.

**Comment by brian-mann** [comment]
With the object literal pattern - its really a "subset" kind of assertion, which means that the key's of the object are what you're asserting against.

If given a blank object then I believe this would give you a specific error telling you to provide **some** kind of assertion otherwise there's nothing specifically we can assert about.

**Comment by jennifer-shehane** [comment]
Closing due to inactivity. We can reopen a new PR if we feel the work is something we want to prioritize in the future.

**Pull request closed without merging**